core/cas-server-core-services/src/test/java/org/apereo/cas/services/AbstractServicesManagerTests.java

```diff
@@ -180,16 +177,91 @@ public void verifyServiceCanBeFoundDuringLoadWithoutCacheInvalidation() throws E
         loadingThread.join();
     }
 
+    /**
+     * Verify that a service which is <strong>not</strong> associated with the current environment ({@code current-env}
+     * in this test) is not incorrectly selected due to "promotion" through the services cache. Scenario is as follows:
+     * <ul>
+     *      <li/> Active environment is {@code current-env}.
+     *      <li/> Service {@code https://prod.test.edu} associated with {@code current-env}, low
+     *      {@code evaluationOrder}.
+     *      <li/> Service {@code https://*.test.edu} <strong>not</strong> associated with {@code current-env}, high
+     *      {@code evaluationOrder}.
+     *      <li/> {@code https://prod.test.edu} is matched correctly <em>if</em> no request for some other domain which
+     *      would match {@code https://*.test.edu} is executed first.
+     *      <li/> {@code https://prod.test.edu} fails to match <em>if</em> a request for some other domain which would
+     *      match {@code https://*.test.edu} is executed first.
+     * </ul>
+     * Failure is caused by "promotion" through the services cache. When the non-{@code https://prod.test.edu} request
+     * is matched, it is added to the cache <em>before</em> its associated environments were evaluated in
+     * {@code validateRegisteredService()}: it ends up in cache even though it doesn't match the current environment.
```

Comment:
Please use a link tag if you are to refer to methods by their name.

Reply:
Can't, it's a private method in a different class. I believe this would break Javadoc compilation.

Reply:
In that case, please remove it from the javadocs.

Reply:
Removed.